versioned/persist/mongodb/src/main/java/org/projectnessie/versioned/persist/mongodb/MongoDatabaseAdapter.java

```diff
@@ -485,4 +521,91 @@ protected RefLog fetchFromRefLog(NonTransactionalOperationContext ctx, Hash refL
       NonTransactionalOperationContext ctx, List<Hash> hashes) {
     return fetchPage(client.getRefLog(), hashes, ProtoSerialization::protoToRefLog);
   }
+
+  @Override
+  protected void writeAttachments(Stream<Entry<AttachmentKey, AttachmentValue>> attachments) {
+    attachments
+        .map(
+            e -> {
+              AttachmentValue value = e.getValue();
+              Document doc = toDoc(attachmentKeyAsString(e.getKey()), value.toByteArray());
+              if (value.hasVersion()) {
+                doc.put(VALUE_VERSION_PROPERTY_NAME, value.getVersion());
+              }
+              return doc;
+            })
+        .forEach(
+            doc -> {
+              if (!client
+                  .getAttachments()
+                  .updateOne(
+                      Filters.eq(doc.get(ID_PROPERTY_NAME)),
+                      new Document("$set", doc),
+                      new UpdateOptions().upsert(true))
+                  .wasAcknowledged()) {
+                throw new IllegalStateException(
+                    "Unacknowledged write to " + client.getAttachments().getNamespace());
+              }
+            });
+  }
+
+  @Override
+  protected boolean consistentWriteAttachment(
+      AttachmentKey key, AttachmentValue value, Optional<String> expectedVersion) {
+    Document doc = toDoc(attachmentKeyAsString(key), value.toByteArray());
+    doc.put(VALUE_VERSION_PROPERTY_NAME, value.getVersion());
+
+    Document id = doc.get(ID_PROPERTY_NAME, Document.class);
+
+    if (expectedVersion.isPresent()) {
+      UpdateResult r =
+          client
+              .getAttachments()
+              .updateOne(
+                  Filters.and(
+                      Filters.eq(id),
+                      Filters.eq(VALUE_VERSION_PROPERTY_NAME, expectedVersion.get())),
+                  new Document("$set", doc));
+      if (!r.wasAcknowledged()) {
+        throw new IllegalStateException(
+            "Unacknowledged write to " + client.getAttachments().getNamespace());
+      }
+      return r.getMatchedCount() == 1L && r.getModifiedCount() == 1L;
+    } else {
```

Comment:
do we want to execute the same else block for `getMatchedCount()>1` and `getMatchedCount()==0`?
Or should those two cases be handled separately?

Reply:
Not sure what you mean here wrt to `return` and the following `else`